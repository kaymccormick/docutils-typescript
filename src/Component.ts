/**
 * @uuid d399ca76-1898-4d75-9b42-b73180707244
 */
import TransformSpec from './TransformSpec';
import { LoggerType } from './types';

/* eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars */
const __docformat__ = 'reStructuredText';

/**
 * Base class for docutils components.
 * Extends TransformSpec, which itself extends SettingsSpec. Thus all
 * docutils components inherit numerous properties and "capabilities"
 * from this inheritance chain. All should probably be combined into
 * a single class since nothing is gained from having the inheritance
 * hierarchy.
 * @extends TransformSpec
 *  
 * @uuid a3d83bd3-67ad-4af3-8167-34d049f1c67c
 */
class Component extends TransformSpec {
    public supported: string[] = [];
    public componentType: string = 'random';
    public toString(): string {
        return `Component<${this.constructor.name}>`;
    }
}


export default Component;