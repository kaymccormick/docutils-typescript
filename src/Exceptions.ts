/**
 * @uuid 2afdc5fd-85c2-4a2f-827c-abe96d2d1076
 */
import { NodeInterface } from "./types";

/*
 * @uuid eb912b7c-0983-4271-863b-23130cecd528
*/
export class UnimplementedError extends Error {
    // @ts-ignore
    public constructor(message?: string) {
        super(message);
        /* instanbul ignore else */
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, UnimplementedError);
        }
        this.message = message || '';
    }
}


/*
 * @uuid b3a69d18-caac-4a4e-9bba-25868c870b5c
*/
export class InvalidStateError extends Error {
    // @ts-ignore
    public constructor(message?: string) {
        super(message);
        /* instanbul ignore else */
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, InvalidStateError);
        }
        this.message = message || '';
    }
}

/*
 * @uuid 432e6258-f2df-4af3-8add-33431b3d5275
*/
export class EOFError extends Error {
    public constructor() {
        super();
        /* instanbul ignore else */
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, EOFError);
        }
    }
}


/*
 * @uuid 2988a140-c2ec-43e7-a293-cb00de925fe6
*/
export class InvalidArgumentsError extends Error {
    public constructor(message: string) {
        super();
        this.message = message;
        /* instanbul ignore else */
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, InvalidArgumentsError);
        }
    }
}

export const InvalidArgumentError = InvalidArgumentsError;


/*
 * @uuid 418c8fbe-01d1-4fbc-892f-8a4397479528
*/
export class SystemMessage extends Error {
    public msg: NodeInterface;
    public level: number;
    public constructor(msg: NodeInterface, level: number, ...params: []) {
        super(...params);
        this.message = msg.astext();
        this.msg = msg;
        this.level = level;
        /* instanbul ignore else */
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, SystemMessage);
        }
    }
}
interface ErrorArgs {
    error?: Error | undefined;
}

/*
 * @uuid f57f1c08-f562-4427-b460-6901eb9a4bdb
*/
export class ApplicationError extends Error {
    public error: Error | undefined;
    public args: ErrorArgs;
    public constructor(message: string, args: ErrorArgs={}) {
        super(message);
        this.args = args;
        if(args !== undefined) {
            this.error = args.error;
        }
        /* instanbul ignore else */
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, ApplicationError);
        }
    }
}

/*
 * @uuid cca9cf0f-f4d9-4514-ad0a-d35d35aea8a0
*/
export class DataError extends ApplicationError {
    public constructor(message: string, args?: ErrorArgs) {
        super(message, args);
        /* instanbul ignore else */
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, DataError);
        }
    }
}

/*
 * @uuid aabf8730-b3d3-4d6e-a550-9ff63bf96dd0
*/
export class AssertError extends Error {
    public constructor(message: string) {
        super(message);
        /* instanbul ignore else */
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, AssertError);
        }
        this.message = message;
    }
}
