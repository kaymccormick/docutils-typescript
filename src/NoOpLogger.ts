import { LoggerType, LeveledLogMethod } from './types';

export class NoOpLogger implements LoggerType {
    debug: LeveledLogMethod;
    silly: LeveledLogMethod;
    error: LeveledLogMethod;
    warn: LeveledLogMethod;
    info: LeveledLogMethod;
    public constructor() {
        const f = (): LoggerType => { return this; };
        //@ts-ignore
        this.debug = f;
        //@ts-ignore
        this.info = f;
        //@ts-ignore
        this.silly = f;
        //@ts-ignore
        this.error = f;
        //@ts-ignore
        this.warn = f;
    }
}
