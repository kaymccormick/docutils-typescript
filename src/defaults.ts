/**
 * @uuid 01ca88e9-b4ab-448d-944d-5af49bb62b08
 */
const defaultReaderName = 'standalone';
const defaultParserName = 'restructuredtext';
const defaultEnableExitStatus = true;

export { defaultReaderName, defaultParserName, defaultEnableExitStatus };
