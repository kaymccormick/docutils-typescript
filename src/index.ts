/**
 *  eslint-disable-next-line import/prefer-default-export 
 * @uuid 62a09a7d-88ab-4fd5-bc03-98b19d2cc427
 */
export { parse } from './parse';
export { StringOutput, StringInput } from './io';
export { StandaloneReader } from './readers/standalone';
export { newDocument } from './newDocument';
import * as nodes from './nodes';
import Writer from './Writer';
import Transform from './Transform';
export { Reader } from './Reader';
export { Publisher } from './Publisher'
export { defaults } from '../gen/defaults';
export { pojoTranslate } from './fn/pojoTranslate';
export { htmlTranslate } from './fn/htmlTranslate';
export { RSTParser } from './parsers/restructuredtext';
export { XMLWriter } from './writers/xml'

export { StringOutput, StringInput } from "./io";
export { StandaloneReader } from "./readers/standalone";
export { newDocument } from "./newDocument";
import * as nodes from "./nodes";
import Writer from "./Writer";
import Transform from "./Transform";
export { Reader } from "./Reader";
export { Publisher } from "./Publisher";
export { defaults } from "../gen/defaults";
export { pojoTranslate } from "./fn/pojoTranslate";
export { htmlTranslate } from "./fn/htmlTranslate";
export { RSTParser } from "./parsers/restructuredtext";
export { XMLWriter } from "./writers/xml";

/* eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars */
const __docformat__ = "reStructuredText";

export const __version__ = "0.14js";
export { nodes, Writer, Transform };