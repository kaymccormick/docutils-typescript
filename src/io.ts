/**
 * @uuid 2b0c8f26-be6e-45a6-951b-96c70f71f55b
 */
import Input from './io/Input';
import Output from './io/Output';
import { ReadInputCallback, LoggerType } from "./types";

/** Direct string input. 
 * @uuid ae2c2dd1-3766-4ae6-a68c-a3a4a0b71cb3
*/
export class StringInput extends Input {
    public constructor(source: string, logger: LoggerType, sourcePath?: string, encoding?: string, errorHandler?: string) {
        super({source, sourcePath, encoding, errorHandler, logger});
        this.sourcePath = '<string>';
    }

    public read(): Promise<any> {
        return Promise.resolve(this.source);
    }
}

/*
 * @uuid 11de15d8-c529-4ddd-830c-6254641c53dd
*/
export class StringOutput extends Output<string> {
    public constructor(
        logger: LoggerType,
        destination?: string,
        destinationPath?: string,
        encoding?: string,
        errorHandler?: string
    ) {
        super(logger, destination, destinationPath, encoding, errorHandler);
        this.defaultDestinationPath = '<string>';

    }

    public write(data: string): string {
        // self.destination = self.encode(data) // fixme encoding
        if (Array.isArray(data)) {
            data = JSON.stringify(data);
        }
        this.destination = data;
        return this.destination;
    }
}

