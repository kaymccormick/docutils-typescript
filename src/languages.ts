/** @uuid 10bfaf36-9321-43dd-8836-1628123c51f8
*/
import * as en from "./languages/en";

import { CoreLanguage, ReporterInterface } from "./types";

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function getLanguage(languageCode: string, reporter: ReporterInterface): CoreLanguage | undefined {
    if (languageCode === "en") {
        return en;
    }

    return undefined;
}