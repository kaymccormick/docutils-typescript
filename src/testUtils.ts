import { newDocument } from './newDocument';
import newReporter    from './newReporter';
import { defaultPublisherOptions } from '../src/constants';
import { getDefaultSettings } from './settingsHelper';
import winston, { format, transports, LoggerOptions, Logger } from 'winston';
import { defaultConsoleLogLevel } from './constants';
import {Publisher} from './Publisher';
import path from 'path';
import StateFactory from './parsers/rst/StateFactory';
import RSTStateMachine from './parsers/rst/RSTStateMachine';
import { LoggerType } from './types';
import process from 'process';

let _logger: LoggerType|undefined;

export function createLogger(options?: LoggerOptions): LoggerType {
    if(_logger !== undefined) {
        return _logger;
    }
    let myOpt = options === undefined ? {} : { ...options};
    if(myOpt.format === undefined) {
        myOpt.format = format.json();
    }
    if(myOpt.transports === undefined) {
        myOpt.transports = [
            new transports.Console({level: defaultConsoleLogLevel}),
            new transports.File({filename: `${path.basename(process.argv[1])}-${process.pid}.log`, level: 'silly'}),
        ];
    }
    _logger = winston.createLogger(myOpt);
    return _logger;
}

export function createPublisher() {
    const publisher = new Publisher({ ...defaultPublisherOptions, logger: createLogger() });
    publisher.setComponents(defaultPublisherOptions.readerName, defaultPublisherOptions.parserName, 'xml');
    return publisher;
}

export function createRSTStateMachine(args: {logger: LoggerType}) {
const sm = new RSTStateMachine({
            stateFactory: new StateFactory({logger: args.logger}),
            initialState: 'Body',
            debugFn: args.logger.debug.bind(args.logger),
            debug: true,
	    logger: args.logger,
        });
	return sm;

}


export function createStateFactory(): StateFactory {
    return new StateFactory({ logger: createLogger() });
}

export function createNewDocument(sourcePath: string='default') {
    return newDocument({sourcePath, logger: createLogger()}, getDefaultSettings());
}
  
export function createNewReporter(sourcePath: string='default') {
    return newReporter({sourcePath}, getDefaultSettings());
}
