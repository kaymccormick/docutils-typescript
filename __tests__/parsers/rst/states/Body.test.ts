import sinon from 'sinon';
import Body from '../../../../src/parsers/rst/states/Body';
import StringList from '../../../../src/StringList';
import RSTStateMachine from '../../../../src/parsers/rst/RSTStateMachine';
import StateFactory from '../../../../src/parsers/rst/StateFactory';
import * as nodes from '../../../../src/nodes';
import { fullyNormalizeName } from "../../../../src/nodeUtils";
import { createRSTStateMachine, createLogger } from '../../../../src/testUtils';


const logger = createLogger();

beforeAll(() => {
    sinon.spy(nodes, 'document');
    sinon.spy(nodes, 'citation');
});
afterEach(() => {
    sinon.restore();
});

beforeEach(() => {
});


function createBody(optSm?: any) {
    const stateMachine = optSm || createRSTStateMachine({logger});
    const body = new Body(stateMachine, true);
    return body;
}

test('Body patterns', () => {
    const body = createBody();
    /* Ensure body state patterns haven't changed. */
    expect(body.patterns).toMatchSnapshot();
});

test('Body constructor',
    () => {
        const body = createBody();
        expect(body).toBeDefined();
    });


// '\\.\\.[ ]+_(?![ ]|$)'
// Regex for reference
// new RegExp(`^(_|(?!_)(\`?)(?![ \`])(.+?)${nonWhitespaceEscapeBefore})
// (?<!(?<!\\x00):)${nonWhitespaceEscapeBefore}[ ]?:([ ]+|$)`),
test.skip('hyperlink_target, no args', () => {
    const body = createBody();
//    expect(() => body.hyperlink_target({})).toThrow();
});

test('explicit hyperlink_target, with arg (malformed)', () => {
    const hyperlinkSource = '.. _myname';
    const rgxp = new RegExp('\\.\\.[ ]+_(?![ ]|$)');
    const body = createBody();
    const match = rgxp.exec(hyperlinkSource);
    // Problem here is that it seeks more input!
//    expect(() => body.hyperlink_target(match)).toThrow();
//    const [[target], blank_finish] = body.hyperlink_target(match);
});


test('explicit citation', () => {
    const source = '.. [myCitation]';
    // fix this in original source
    // we extract the regexp so we can pass in the proper results
    const rgxp = new RegExp('\\.\\.[ ]+\\[(\\w+)\\]([ ]+|$)');
    const sm = createRSTStateMachine({logger});
    const body = createBody(sm);
    console.log(fullyNormalizeName);
    //    console.log(sm);
    //    console.log(sm.getSourceAndLine());

    const match = rgxp.exec(source);
    if(match) {
    body.citation(match);
    }
//    expect(() => body.citation(match)).toThrow();
//    const [[target], blank_finish] = body.hyperlink_target(match);
});

