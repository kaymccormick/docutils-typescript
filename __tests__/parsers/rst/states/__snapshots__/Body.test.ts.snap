// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body patterns 1`] = `
Object {
  "anonymous": /\\^__\\( \\+\\|\\)/,
  "bullet": /\\^\\[-\\+\\*\\\\u2022\\\\u2023\\\\u2043\\]\\( \\+\\|\\$\\)/,
  "doctest": /\\^>>>\\( \\+\\|\\$\\)/,
  "enumerator": /\\^\\(\\(\\\\\\(\\^\\(\\[0-9\\]\\+\\|\\[a-z\\]\\|\\[A-Z\\]\\|\\[ivxlcdm\\]\\+\\|\\[IVXLCDM\\]\\+\\|#\\)\\\\\\)\\)\\|\\(\\^\\(\\[0-9\\]\\+\\|\\[a-z\\]\\|\\[A-Z\\]\\|\\[ivxlcdm\\]\\+\\|\\[IVXLCDM\\]\\+\\|#\\)\\\\\\)\\)\\|\\(\\^\\(\\[0-9\\]\\+\\|\\[a-z\\]\\|\\[A-Z\\]\\|\\[ivxlcdm\\]\\+\\|\\[IVXLCDM\\]\\+\\|#\\)\\\\\\.\\)\\)\\( \\+\\|\\$\\)/,
  "explicit_markup": /\\^\\\\\\.\\\\\\.\\( \\+\\|\\$\\)/,
  "field_marker": /\\^:\\(\\?!\\[: \\]\\)\\(\\[\\^:\\\\\\\\\\]\\|\\\\\\\\\\.\\|:\\(\\?!\\(\\[ \`\\]\\|\\$\\)\\)\\)\\*\\(\\?<! \\):\\( \\+\\|\\$\\)/,
  "grid_table_top": /\\\\\\+-\\[-\\+\\]\\+-\\\\\\+ \\*\\$/,
  "line": /\\^\\(\\[!-\\\\/:-@\\[-\`\\{-~\\]\\)\\\\1\\* \\*\\$/,
  "line_block": /\\^\\\\\\|\\( \\+\\|\\$\\)/,
  "option_marker": /\\^\\(\\(-\\|\\\\\\+\\)\\[a-zA-Z0-9\\]\\( \\?\\(\\[a-zA-Z\\]\\[a-zA-Z0-9_-\\]\\*\\|<\\[\\^<>\\]\\+>\\)\\)\\?\\|\\(--\\|\\\\/\\)\\[a-zA-Z0-9\\]\\[a-zA-Z0-9_-\\]\\*\\(\\[ =\\]\\(\\[a-zA-Z\\]\\[a-zA-Z0-9_-\\]\\*\\|<\\[\\^<>\\]\\+>\\)\\)\\?\\)\\(, \\(\\(-\\|\\\\\\+\\)\\[a-zA-Z0-9\\]\\( \\?\\(\\[a-zA-Z\\]\\[a-zA-Z0-9_-\\]\\*\\|<\\[\\^<>\\]\\+>\\)\\)\\?\\|\\(--\\|\\\\/\\)\\[a-zA-Z0-9\\]\\[a-zA-Z0-9_-\\]\\*\\(\\[ =\\]\\(\\[a-zA-Z\\]\\[a-zA-Z0-9_-\\]\\*\\|<\\[\\^<>\\]\\+>\\)\\)\\?\\)\\)\\*\\(  \\+\\| \\?\\$\\)/,
  "simple_table_top": /=\\+\\( \\+=\\+\\)\\+ \\*\\$/,
  "text": /\\(\\?:\\)/,
}
`;
